### DISCUSSION TITLE
How to delete CKAN activity history?

### DISCUSSION THREAD

**Discussion by themowski** [ckan/ckan#8187] [Q&A] [answered]
Our team is using CKAN 2.10.1 with the activity plugin. We have recently done some data migrations (e.g., making each dataset's `name`/URL match its `id`, modifying our `ckanext-scheming` schemas), and we noticed some breakage in our datasets' Activity Stream pages as a result -- for example, clicking on "View this version" links results in an internal server error, and the logs show a super long traceback with errors that span the full templating pipeline.[^1]

While we would like the preserve the activity, we don't feel that it's worth the time required to debug & potentially wait for upstream patches -- deleting the activity is good enough for us, because it would eliminate the links that are causing issues, and we've verified that "new" activity (from after our data migrations) works fine.

I couldn't find any documentation about how to safely delete activity, and I don't see any options for it in the CKAN UI, so I thought I'd ask here. My naive approach to delete activity is the following:

* Use `psql` to connect to the `ckandb` database in PostgreSQL as the appropriate admin user
* Run `delete from activity;` to delete all rows in the `activity` table
* Run `reindex database ckandb;` to fully reindex the database

I noticed there is also a table called `activity_detail`, but it is empty, so I left it alone.

After doing this on a test instance of CKAN, I verified that there were no obvious ill effects.

The big question is, are these steps sufficient? Are they safe? Any further guidance would be appreciated.

[^1]: My intuition based on the traceback is that this is related to all of the schema changes we've done recently, including introducing new Organization types and moving existing datasets into the new Organizations, but that's not really relevant to the question at hand.

**Comment by wardi** [comment, marked as answer]
We *really* should have a command to do this in core, but for now you can implement this yourself something like this https://github.com/open-data/ckanext-canada/blob/c2808d9c318e19d5f0c3b4c2f53073e9576d6b8f/ckanext/canada/commands.py#L740-L756

**Reply by themowski** [reply]
I definitely agree that having a built-in CLI command would be useful, that was the first place I checked in my search for this functionality.

It looks like your code snippet does the same thing as my raw SQL commands (except that I wipe all history, whereas your code preserves some number of days), so pending a quick investigation to make sure that they are indeed identical, it looks like that's my answer. We'll either add something like this to our plugin, or just fall back to the raw SQL.

Many thanks as usual for the quick & helpful response! We really appreciate the support that you & the other CKAN devs have given us. I know it's a thankless job most days, but know that it's appreciated. 😃
